keps/sig-node/127-usernamespaces-support/README.md

```diff
@@ -0,0 +1,842 @@
+  container and in the host. This mode is intended for pods sharing volumes as
+  they will run with the same effective IDs on the host, allowing them to read
+  and write files in the shared storage.
+
+- **Pod**:
+  Each pod is placed in a different user namespace and has a different and
+  non-overlapping ID mapping across the node. This mode is intended for stateless pods, i.e.
+  pods using only ephemeral volumes like `configMap,` `downwardAPI`, `secret`,
+  `projected` and `emptyDir`. This mode not only provides host-to-pod isolation
+  but also pod-to-pod isolation as each pod has a different range of effective
+  IDs in the host.
+
+### User Stories
+
+#### Story 1
+
+As a cluster admin, I want run some pods with privileged capabilities because
+the applications in the pods require it (e.g. `CAP_SYS_ADMIN` to mount a FUSE
+filesystem or `CAP_NET_ADMIN` to setup a VPN) but I don't want this extra
+capability to grant me any extra privilege on the host.
+
+#### Story 2
+
+As a cluster admin, I want to allow some pods to run in the host user namespace
+if they need a feature only available in such user namespace, such as loading a
+kernel module with `CAP_SYS_MODULE`.
+
+### Notes/Constraints/Caveats
+
+#### Volumes Support
+
+The Linux kernel uses the effective user and group IDs (the ones the host) to
+check the file access permissions. Since with user namespaces IDs are mapped to
+a different value on the host, this causes issues accessing volumes if the pod
+is run with a different mapping, i.e. the effective user and group IDs on the
+host change.
+
+This proposal supports volumes without changing the user and group IDs and
+leaves that problem for the user to manage. Developing Linux kernel features
+such as shiftfs could allow different pods to see a volume with its own IDs but
+it is not available yet. Among the possible future kernel solutions, we can
+list:
+
+- [shiftfs: uid/gid shifting filesystem](https://lwn.net/Articles/757650/)
+- [A new API for mounting filesystems](https://lwn.net/Articles/753473/)
+- [user_namespace: introduce fsid mappings](https://lwn.net/Articles/812221/)
+
+In regard to this proposal, volumes can be divided in ephemeral and
+non-ephemeral.
+
+Ephemeral volumes are associated to a **single** pod and their lifecyle is
+dependent on that pod. These are `configMap`, `secret`, `emptyDir`,
+`downwardAPI`, etc. These kind of volumes can work with any of the three
+different modes of `userNamespaceMode` as they are not shared by different pods
+and hence all the processes accessing those volumes have the same effective user
+and group IDs. Kubelet creates the files for those volumes and it can easily set
+the file ownership too.
+
+Non-ephemeral volumes are more difficult to support since they can be persistent
+and shared by multiple pods. This proposal supports volumes with two different
+strategies:
+- The `Cluster` mode makes it easier for pods to share files using volumes when
```

Comment:
Is SA the right semantics though? Easy to imagine shared access to one resource by multiple principals, generally speaking. Might still be the convenient choice.

Reply:
Problem space feels something like same-origin on the Web, K8s namespace might be the natural analogue...

Also worth thinking about cases where users have two clusters with the same set of namespaces and expect to move workloads between them - could we ensure that the mapping was consistently configured for both clusters, even across concurrent, potentially out-of-order additions/removals of namespaces?

Reply:
Perhaps we could derive the mapping from the namespace name, to guarantee it's the same in different clusters. However I don't know the security implications of allowing users to indirectly control the ranges by the namespace name. 

Cases like this are another reason why we're want to have the same mapping for all pods with volumes.

Reply:
If we started with something like Cluster mode (which I think is a great first step), it theoretically blocks us from transparently migrating to Namespace mode in the future, since mappings would likely change, right? At that point do we basically have to wait for shiftfs to make things more secure without additional user intervention?

Reply:
> I don't know the security implications of allowing users to indirectly control the ranges by the namespace name.

It _might_ not be a huge problem, given namespaces are probably created by cluster admins, controlling ranges might fit their privilege level anyway.

Reply:
> If we started with something like Cluster mode (which I think is a great first step), it theoretically blocks us from transparently migrating to Namespace mode in the future, since mappings would likely change, right? 

Yes, users will have to update the volumes ownership. Btw, one of our goals with this proposal is to have something that works good for now and that allows to easily use a shiftfs-like solution when it'll be available. In one of our comments (https://github.com/kubernetes/enhancements/pull/2101#issuecomment-754778887) we're proposing to have a "Cluster" mode but implicit, to avoid changes to the pod spec. 

> At that point do we basically have to wait for shiftfs to make things more secure without additional user intervention?

Yes, I think shiftfs (or a similar solution) is the only way to avoid user intervention at all.

Reply:
Ok, thanks for the clarification.